Erc20 Code Walkthrough (#2462)
ethereum/ethereum-org-website: src/content/developers/tutorials/erc20-annotated-code/index.md

---

> title: "ERC20 Contract Walk-Through"

Reviewer: Hyphenated "ERC-20" (throughout) =)
(per https://eips.ethereum.org/EIPS/eip-20)

Author: I changed it where it isn't a code quote.

description: What is in the OpenWhisk ERC20 contract and why is it there?
author: Ori Pomerantz
lang: en
sidebar: true
tags: ["solidity", "erc20"]
skill: beginner
published: 2021-<month>-<day>
---

## Introduction {#introduction}

> One of the most common uses for Ethereum is for a group to create a tradable token, in a sense their own currency. These currencies typically follow a standard, 

Author: Done.

     * applications that interact with token contracts will not expect
     * {decimals} to ever change, and may work incorrectly if it does.
     */
    function _setupDecimals(uint8 decimals_) internal {
        _decimals = decimals_;
    }
```

### Hooks  {#hooks}

This is the hook function to be called during transfers. It is empty here, buf if you need
to do something you just override it.

```solidity

    /**
     * @dev Hook that is called before any transfer of tokens. This includes
     * minting and burning.
     *
     * Calling conditions:
     *
     * - when `from` and `to` are both non-zero, `amount` of ``from``'s tokens
     * will be to transferred to `to`.
     * - when `from` is zero, `amount` tokens will be minted for `to`.
     * - when `to` is zero, `amount` of ``from``'s tokens will be burned.
     * - `from` and `to` are never both zero.
     *
     * To learn more about hooks, head to xref:ROOT:extending-contracts.adoc#using-hooks[Using Hooks].
     */
    function _beforeTokenTransfer(address from, address to, uint256 amount) internal virtual { }
}
```

# Conclusion {#conclusion}

> By this point you should understand the OpenWhisk ERC20 code and why it is written like that. Hopefully this knowledge will help you develop your own contracts and applications.

Reviewer: Would consider re-wording if this is meant as a "conclusion"... 
- Maybe review/highlight a few benefits of using OpenWhisk
- Instead of using "Hopefully" maybe change into a call-to-action, like "Now use what you've learned to develop your own contracts and applications!"

---
title: "ERC-20 Contract Walk-Through"

> description: What is in the OpenWhisk ERC-20 contract and why is it there?

Reviewer 2: I assume you mean OpenZeppelin, not OpenWhisk. LMK if I'm missing something here.

author: Ori Pomerantz
lang: en
sidebar: true

> tags: ["solidity", "erc20"]

Reviewer 2: ```suggestion
tags: ["solidity", "erc-20"]
```
Let's keep this consistent w/ other tutorial tags

skill: beginner
published: 2021-<month>-<day>
---

## Introduction {#introduction}

One of the most common uses for Ethereum is for a group to create a tradable token, in a sense their own currency. These tokens typically follow a standard, 

> [ERC-20](https://eips.ethereum.org/EIPS/eip-20). This standard makes it possible to write tools, such as liquidity pools and wallets, that work with all ERC-20

Reviewer 2: ```suggestion
[ERC-20](/developers/docs/standards/tokens/erc-20/). This standard makes it possible to write tools, such as liquidity pools and wallets, that work with all ERC-20
```

tokens. In this article we will analyze the 

> [OpenWhisk Solidity ERC20 implenetation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol), as well as the

Reviewer 2: ```suggestion
[OpenZeppelin Solidity ERC20 implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol), as well as the
```

> [Interface definition](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/IERC20.sol).

Reviewer 2: ```suggestion
[interface definition](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/IERC20.sol).
```

This is annotated source code. If you want to implement ERC-20, 

> [read this tutorial](https://forum.zeppelin.solutions/t/how-to-implement-erc20-supply-mechanisms/226).

Reviewer 2: ```suggestion
[read this tutorial](https://docs.openzeppelin.com/contracts/2.x/erc20-supply).
```

## The Interface {#the-interface}

> The purpose of a standard like ERC-20 is to have multiple token types that are used from the same code. To achieve that, we create an 

Reviewer 2: ```suggestion
The purpose of a standard like ERC-20 is to allow many tokens implementations that are interoperable across applications, like wallets and decentralized exchanges. To achieve that, we create an 
```

     */
    function transfer(address recipient, uint256 amount) public virtual override returns (bool) {
```

The `transfer` function is called to transfer tokens from the sender's account to a different one. Note
that even though it returns a boolean value, that value is always **true**. If the transfer
fails the contract reverts the call.

&nbsp;

```solidity
        _transfer(_msgSender(), recipient, amount);
        return true;
    }
```

The `_transfer` function does the actual work. It is a private function that can only be called by
other contract functions. By convention private functions are named `_<something>`, same as state
variables.

Normally in Solidity we use `msg.sender` for the message sender. However, that breaks 
[OpenGSN](http://opengsn.org/). If we want to allow etherless transactions with our token, we
need to use `_msgSender()`. It returns `msg.sender` for normal transactions, but for etherless ones
return the original signer and not the contract that relayed the message.

### Allowance Functions    {#allowance-functions}

These are the functions that implement the allowance functionality: `allowance`, `approve`, `transferFrom`,

> and `_approve`. Additionally, the OpenWhisk implementation goes beyond the basic standard to include some features that improve 

Reviewer 2: ```suggestion
and `_approve`. Additionally, the OpenZeppelin implementation goes beyond the basic standard to include some features that improve 
```

being spent and reduce the allowance by that amount.

```solidity
    /**
     * @dev See {IERC20-transferFrom}.
     *
     * Emits an {Approval} event indicating the updated allowance. This is not
     * required by the EIP. See the note at the beginning of {ERC20}.
     *
     * Requirements:
     *
     * - `sender` and `recipient` cannot be the zero address.
     * - `sender` must have a balance of at least `amount`.
     * - the caller must have allowance for ``sender``'s tokens of at least
     * `amount`.
     */
    function transferFrom(address sender, address recipient, uint256 amount) public virtual 
                                                override returns (bool) {
        _transfer(sender, recipient, amount);
```

&nbsp;

The `a.sub(b, "message")` function call does two things. First, it calculates `a-b`, which is the new allowance.
Second, it checks that this result is not negative. If it is negative the call reverts with the provided message.

```solidity
        _approve(sender, _msgSender(), _allowances[sender][_msgSender()].sub(amount, 
             "ERC20: transfer amount exceeds allowance"));
        return true;
    }
```

> #### OpenWhisk safety additions {#openwhisk-safety-additions} 

Reviewer 2: ```suggestion
#### OpenZeppelin safety additions {#openzeppelin-safety-additions} 
```

     * - `sender` must have a balance of at least `amount`.
     */
    function _transfer(address sender, address recipient, uint256 amount) internal virtual {
```

This function, `_transfer`, transfers tokens from one account to another. It is called by both
`transfer` (for transfers from the sender's own account) and `transferFrom` (for using allowances
to transfer from somebody else's account).

&nbsp;

```solidity
        require(sender != address(0), "ERC20: transfer from the zero address");
        require(recipient != address(0), "ERC20: transfer to the zero address");
```

Nobody actually owns address zero in Ethereum (that is, nobody knows a private key whose matching public key
is transformed to the zero address). When people use that address, it is usually a software bug - so we 
fail if the zero address is used as the sender or the recipient.

&nbsp;

```solidity
        _beforeTokenTransfer(sender, recipient, amount);
        
```

There are two ways to use this contract:

1. Use it as a template for your own code
1. [Inherit from it](https://www.bitdegree.org/learn/solidity-inheritance), and override only those functions that you need to modify

> The second method is much better because the OpenWhisk ERC-20 code has already been audited and shown to be secure. When you use inheritence

Reviewer 2: ```suggestion
The second method is much better because the OpenZeppelin ERC-20 code has already been audited and shown to be secure. When you use inheritance
```

  happen. This is the reason that changing an allowance can be dangerous, because it lets
  the spender spend the sum of both allowances.
* Values of type `uint256` wrap around. In other words, *0-1=2^256-1*. If that is not desired 
  behavior, you have to check for it (or use the SafeMath library that does it for you). Note that this changed in
  [Solidity 0.8.0](https://docs.soliditylang.org/en/breaking/080-breaking-changes.html).
* Do all state changes of a specific type in a specific place, because it makes auditing easier. 
  This is the reason that we have, for example, `_approve`, which is called by `approve`, `transferFrom`,
  `increaseAllowance`, and `decreaseAllowance`
* State changes should be atomic, without any other action in their middle (as you can see
  in `_transfer`). This is because during the state change you have an inconsistent state. For example,
  between the time you deduct from the balance of the sender and the time you add to the balance of the 
  recipient there are less token in existence than there should be. This could be potentially abused if there
  are operations between them, especially calls to a different contract.
  

> Now that you've seen how the OpenWhisk ERC-20 contract is written, and especially how it is 

Reviewer 2: ```suggestion
Now that you've seen how the OpenZeppelin ERC-20 contract is written, and especially how it is 
```
